File: unitymodule/src/main/java/ir/firoozeh/unitymodule/Handlers/UnityLogin.java
package ir.firoozeh.unitymodule.Handlers;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.os.IBinder;
import android.util.Log;

import ir.firoozeh.gameservice.IAsyncGameServiceCallback;
import ir.firoozeh.gameservice.ILoginGameServiceInterface;
import ir.firoozeh.unitymodule.Interfaces.IGameServiceCallback;
import ir.firoozeh.unitymodule.Interfaces.IGameServiceLoginCheck;
import ir.firoozeh.unitymodule.Interfaces.InstallDialogListener;
import ir.firoozeh.unitymodule.Interfaces.UpdateDialogListener;
import ir.firoozeh.unitymodule.Interfaces.UpdateUtilListener;
import ir.firoozeh.unitymodule.Utils.ConnectivityUtil;
import ir.firoozeh.unitymodule.Utils.DialogUtil;
import ir.firoozeh.unitymodule.Utils.UpdateUtil;

/**
 * Created by Alireza Ghodrati on 03/10/2018 in
 * Time 11:34 AM for App ir.firoozeh.unitymodule
 */

public final class UnityLogin implements InstallDialogListener, UpdateDialogListener {

    private static UnityLogin Instance;
    private final String TAG = getClass().getName();
    private ILoginGameServiceInterface loginGameServiceInterface;
    private LoginService loginTestService;
    private Context context;
    private Activity activity;
    private IGameServiceCallback loginCheck;
    private boolean CheckAppStatus = true;
    private boolean CheckOptionalUpdate = true;


    public UnityLogin () {
        Instance = this;
    }

    public static UnityLogin GetInstance () {
        if (Instance == null) Instance = new UnityLogin();
        return Instance;

    }

    public void SetUnityContext (Activity activity) {
        this.context = activity.getApplicationContext();
        this.activity = activity;
    }

    public void InitLoginService (
            boolean CheckAppStatus
            , boolean CheckOptionalUpdate
            , IGameServiceCallback callback) {

        this.loginCheck = callback;
        this.CheckAppStatus = CheckAppStatus;
        this.CheckOptionalUpdate = CheckOptionalUpdate;
        initLoginService(callback);

    }

    public void DisposeService () {
        releaseLoginService();
    }

    private void initLoginService (final IGameServiceCallback callback) {

        try {
            if (isPackageInstalled(context.getPackageManager())) {
                if (ConnectivityUtil.isNetworkConnected(activity)) {
                    int VerCode = getGameServiceVersionCode(context.getPackageManager());
                    if (VerCode != -1) {
                        UpdateUtil.CheckUpdate(CheckOptionalUpdate, VerCode, new UpdateUtilListener() {
                            @Override
                            public void onUpdateAvailable (String ChangeLog, boolean MustUpdate) {
                                DialogUtil.ShowUpdateAppDialog(activity, MustUpdate, ChangeLog, UnityLogin.this);
                            }

                            @Override
                            public void onHaveLastVersion () {
                                _initLoginService(callback);
                            }

                            @Override
                            public void onForceInit () {
                                _initLoginService(callback);
                            }
                        });
                    }
                } else callback.OnError("NetworkUnreachable");
            } else {
                if (CheckAppStatus)
                    DialogUtil.ShowInstallAppDialog(activity, this);
                else
                    callback.OnError("GameServiceNotInstalled");
            }
        } catch (Exception e) {
            Log.e(getClass().getName(), e.toString());
            callback.OnError("GameServiceException");
        }

    }


    private void _initLoginService (final IGameServiceCallback callback) {
        loginTestService = new LoginService();
        Intent i = new Intent(
                "ir.firoozeh.gameservice.Services.LoginService.BIND");
        i.setPackage("ir.firoozeh.gameservice");
        boolean ret = context.bindService(i, loginTestService, Context.BIND_AUTO_CREATE);

        if (!ret)
            callback.OnError("GameServiceNotBounded");
    }

    private void releaseLoginService () {
        context.unbindService(loginTestService);
        loginTestService = null;
        Log.e(TAG, "releaseLoginService(): unbound.");
    }

    public void IsUserLoggedIn (IGameServiceLoginCheck check) {
        try {
            if (loginGameServiceInterface != null)
                check.isLoggedIn(loginGameServiceInterface.isLoggedIn());
            else check.OnError("UnreachableService");
        } catch (Exception e) {
            check.OnError("GameServiceException");
        }
    }

    public void ShowLoginUI (final IGameServiceLoginCheck check) {
        try {
            if (loginGameServiceInterface != null) {
                IAsyncGameServiceCallback.Stub iAsyncGameServiceCallback = new IAsyncGameServiceCallback.Stub() {
                    @Override
                    public void OnCallback (String Result) {
                        check.isLoggedIn(true);
                    }

                    @Override
                    public void OnError (String Error) {
                        check.OnError(Error);
                    }
                };
                loginGameServiceInterface.ShowLoginUI(iAsyncGameServiceCallback);
            } else check.OnError("UnreachableService");
        } catch (Exception e) {
            Log.e(getClass().getName(), e.toString());
            check.OnError("GameServiceException");
        }
    }

    private boolean isPackageInstalled (PackageManager packageManager) {
        try {
            packageManager.getPackageInfo("ir.firoozeh.gameservice", 0);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    private int getGameServiceVersionCode (PackageManager packageManager) {
        try {
            return packageManager.getPackageInfo("ir.firoozeh.gameservice", 0).versionCode;
        } catch (Exception e) {
            return -1;
        }
    }


    @Override
    public void onDismiss () {
        loginCheck.OnError("GameServiceInstallDialogDismiss");
    }

    @Override
    public void onUpdateDismiss () {
        loginCheck.OnError("GameServiceUpdateDialogDismiss");
    }


    private class LoginService
            implements ServiceConnection {
        public void onServiceConnected (ComponentName name, IBinder boundService) {
            loginGameServiceInterface = ILoginGameServiceInterface.Stub
                    .asInterface(boundService);
            loginCheck.OnCallback("Success");
            Log.e(TAG, "LoginTestService(): Connected");
        }

        public void onServiceDisconnected (ComponentName name) {
            loginGameServiceInterface = null;
            Log.e(TAG, "LoginTestService(): Disconnected");
        }
    }

}

File: unitymodule/src/main/java/ir/firoozeh/unitymodule/Handlers/UnityLogin.kt
package ir.firoozeh.unitymodule.Handlers

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.os.IBinder
import android.os.RemoteException
import android.util.Log

import ir.firoozeh.gameservice.IAsyncGameServiceCallback
import ir.firoozeh.gameservice.ILoginGameServiceInterface
import ir.firoozeh.unitymodule.Interfaces.IGameServiceCallback
import ir.firoozeh.unitymodule.Interfaces.IGameServiceLoginCheck

/**
 * Created by Alireza Ghodrati on 03/10/2018 in
 * Time 11:34 AM for App ir.firoozeh.unitymodule
 */

class UnityLogin {

    private val TAG = javaClass.getName()
    private var loginGameServiceInterface: ILoginGameServiceInterface? = null
    private var loginTestService: LoginService? = null
    private var context: Context? = null
    private var loginCheck: IGameServiceCallback? = null

    init {
        Instance = this
    }

    fun SetUnityContext(activity: Activity) {
        this.context = activity.applicationContext
    }

    fun InitLoginService(callback: IGameServiceCallback) {
        this.loginCheck = callback
        initLoginService(callback)
    }

    fun DisposeService() {
        releaseLoginService()
    }

    private fun initLoginService(callback: IGameServiceCallback) {

        try {
            if (isPackageInstalled("ir.firoozeh.gameservice", context!!.packageManager)) {
                loginTestService = LoginService()
                val i = Intent(
                        "ir.firoozeh.gameservice.Services.LoginService.BIND")
                i.setPackage("ir.firoozeh.gameservice")
                val ret = context?.bindService(i, loginTestService, Context.BIND_AUTO_CREATE)

                if (!ret!!)
                    callback.OnError("GameServiceNotBounded")

            } else
                callback.OnError("GameServiceNotInstalled")
        } catch (e: Exception) {
            Log.e(javaClass.name, e.toString())
            callback.OnError("GameServiceException")
        }

    }

    private fun releaseLoginService() {
        context!!.unbindService(loginTestService!!)
        loginTestService = null
        Log.e(TAG, "releaseLoginService(): unbound.")
    }

    fun IsUserLoggedIn(check: IGameServiceLoginCheck) {
        try {
            if (loginGameServiceInterface != null)
                check.isLoggedIn(loginGameServiceInterface?.isLoggedIn!!)
            else
                check.OnError("UnreachableService")
        } catch (e: Exception) {
            check.OnError("GameServiceException")
        }

    }

    fun ShowLoginUI(check: IGameServiceLoginCheck) {
        try {
            if (loginGameServiceInterface != null) {
                val iAsyncGameServiceCallback = object : IAsyncGameServiceCallback.Stub() {
                    @Throws(RemoteException::class)
                    override fun OnCallback(Result: String) {
                        check.isLoggedIn(true)
                    }

                    @Throws(RemoteException::class)
                    override fun OnError(Error: String) {
                        check.OnError(Error)
                    }
                }
                loginGameServiceInterface?.ShowLoginUI(iAsyncGameServiceCallback)
            } else
                check.OnError("UnreachableService")
        } catch (e: Exception) {
            Log.e(javaClass.name, e.toString())
            check.OnError("GameServiceException")
        }

    }

    private fun isPackageInstalled(packagename: String, packageManager: PackageManager): Boolean {
        try {
            packageManager.getPackageInfo(packagename, 0)
            return true
        } catch (e: Exception) {
            return false
        }

    }

    private inner class LoginService : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, boundService: IBinder) {
            loginGameServiceInterface = ILoginGameServiceInterface.Stub
                    .asInterface(boundService)
            loginCheck?.OnCallback("Success")
            Log.e(TAG, "LoginTestService(): Connected")
        }

        override fun onServiceDisconnected(name: ComponentName) {
            loginGameServiceInterface = null
            Log.e(TAG, "LoginTestService(): Disconnected")
        }
    }

    companion object {

        @SuppressLint("StaticFieldLeak")
        private var Instance: UnityLogin? = null

        fun GetInstance(): UnityLogin {
            if (Instance == null) Instance = UnityLogin()
            return Instance as UnityLogin

        }
    }

}
